IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/ApiProtocol.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  ApiProtocol.swift
+//  SwiftSeedProject
+//
+//  Created by Brian Sztamfater on 3/4/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+import Alamofire
+import SwiftyJSON
+
+protocol ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]?, headers: [String : String]?) -> ApiRequestProtocol
+}
+
+extension ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint) -> ApiRequestProtocol {
+        return apiRequest(endpoint: endpoint, parameters: nil, headers: nil)
+    }
+    
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]?) -> ApiRequestProtocol {
+        return apiRequest(endpoint: endpoint, parameters: parameters, headers: nil)
+    }
+}
+
+protocol ApiRequestProtocol {
+    func apiResponse(completionHandler: @escaping (DataResponse<JSON>) -> Void) -> Self
+}
+
+func += <K, V> ( left: inout [K : V], right: [K : V]) {
+    for (k, v) in right {
+        left.updateValue(v, forKey: k)
+    }
+}
+
+extension SessionManager: ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]? = nil, headers: [String : String]? = nil) -> ApiRequestProtocol {
+        // Insert your common headers here, for example, authorization token or accept.
+        var commonHeaders = ["Accept" : "application/json"]
+        if let headers = headers {
+            commonHeaders += headers
+        }
+        
+        return request(endpoint.url, method: endpoint.method, parameters: parameters, encoding: JSONEncoding.default, headers: commonHeaders)
+    }
+}
+
+extension DataRequest: ApiRequestProtocol {
+    static func apiResponseSerializer() -> DataResponseSerializer<JSON> {
+        return DataResponseSerializer<JSON> { request, response, data, error in
+            if let error = error {
+                return .failure(error)
+            }
+            
+            guard let validData = data else {
+                let domain = Bundle.main.infoDictionary?[kCFBundleNameKey as String] as? String
+                let reason = "Data could not be serialized. Input data was nil."
+                return .failure(NSError(domain: domain!, code: 1001, userInfo: [NSLocalizedDescriptionKey : reason]))
+            }
+            
+            let json = JSON(data: validData)
+            // TODO: Should consider HTTP response as well.
+            return sanitizeError(json: json)
+        }
+    }
+    
+    static func sanitizeError(json: JSON) -> Result<JSON> {
+        if !json["error"].exists() {
+            return .success(json)
+        }
+        
+        let code = json["error"]["code"].intValue
```

Comment:
From the perspective of a reusable class, this code is trouble.

You're assuming error messages from the API will have this specific format.  Even the same application can consume two different APIs with different error format.